pkg/kubelet/pluginmanager/pluginwatcher/example_handler.go

```diff
@@ -26,8 +26,8 @@ import (
 	"golang.org/x/net/context"
 	"k8s.io/klog"
 
-	v1beta1 "k8s.io/kubernetes/pkg/kubelet/util/pluginwatcher/example_plugin_apis/v1beta1"
-	v1beta2 "k8s.io/kubernetes/pkg/kubelet/util/pluginwatcher/example_plugin_apis/v1beta2"
+	v1beta1 "k8s.io/kubernetes/pkg/kubelet/pluginmanager/pluginwatcher/example_plugin_apis/v1beta1"
+	v1beta2 "k8s.io/kubernetes/pkg/kubelet/pluginmanager/pluginwatcher/example_plugin_apis/v1beta2"
 )
```

Comment:
Update the "gRPC Service Lifecycle" section of `pluginmanager/pluginwatcher/README.md` file to indicate that if any of the steps in registration fails, on retry registration will start from scratch.
For example, imagine for a new socket:

* Registration.GetInfo is called against socket. Succeeds.
* Validate is called against internal plugin type handler. Succeeds.
* Register is called against internal plugin type handler. Fails.
* NotifyRegistrationStatus is called against socket to indicate registration failed. Succeeds.

After some back off time time, the `pluginmanager` will try again from the begining:

* GetInfo is called against socket, again. Succeeds.
* Validate is called against internal plugin type handler again. Succeeds.
* Register is called against internal plugin type handler again. Succeeds this time.
* NotifyRegistrationStatus is called against socket to indicate registration succeed. Succeeds.
* Plugin is successfully registered.

May also be a good to document what happens if for `GetInfo` call to socket returns a different "type", "endpoint" or "supported version" then a previous successful registration for the same "socket path" in the readme.

Reply:
@saad-ali I'm finishing off the last bit of changes but this will no longer be true:
```
what happens if for GetInfo call to socket returns a different "type", "endpoint" or 
"supported version" then a previous successful registration for the same 
"socket path"
```
Because `GetInfo` is now called in the register and unregister calls (`GenerateRegisterPluginFunc` and `GenerateUnregisterPluginFunc`), I've updated the actual/desired state caches so they no longer keep information about plugins except for the socket path and `foundInDeprecatedDir` (I haven't pushed the changes yet). As a result, we won't be able to check if a plugin exists for the same socket path but other information has changed. I think this is okay because `plugin_watcher` only trigger plugin creation when a new file is detected, so the scenario of plugin information changing for an existing socket file won't happen since I'm leaving the [`Start()` method](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/util/pluginwatcher/plugin_watcher.go#L89) untouched

Reply:
The problem is that if a file is deleted and recreated very quickly, the reconciler may not even notice that it was removed from the desired state. So reregistration won't happen.

@msau42 had a great suggestion for how we can still retain the cool reregisteration behavior:

1. Generate and store current timestamp during `desiredStateOfWorld.AddPlugin(...)` to the `desiredStateOfWorld` struct.
2. Upon successful registration during `actualStateOfWorld.AddPlugin(...)` copy the timestamp from `desiredStateOfWorld` to the `actualStateOfWorld` struct.
3. Have `desiredStateOfWorld.PluginExists(...)` verify that the timestamps in desired and actual state match, if not return false.

This way, after a successful registration, when the socket is deleted and recreated, it will force a reregistration because the timestamp of the socket in the desired state will be newer then the timestamp of the socket in the actual state. Specifically reconciler will first check for volume plugins to unregister, it will run `desiredStateOfWorld.PluginExists` which will return false. This will start a unregistration operation. Once the unregsitration is complete the socket will be removed from actual state. The reconciler will see the desired state has a socket that actual state doesn't and will start a registration operation.

Reply:
@saad-ali @msau42 how do we determine whether to add a plugin to the desired state of world cache when traversing directories in `plugin_watcher`? In the case when the socket is deleted (and the socket is not removed from the desired state cache because it happens very quickly) and recreated, if we do what you suggested in number 3

> Have `desiredStateOfWorld.PluginExists(...)` verify that the timestamps in desired and actual state match, if not return false.

when we are traversing directories in `plugin_watcher`, in `handlePluginRegistration(...)` `desiredStateOfWorld.PluginExists(...)` will return true because the timestamps in desired and actual state match, and a reregistration won't eventually happen.

Reply:
@taragu, I forgot that `pluginwatcher` called `desiredStateOfWorld.PluginExists(...)` before calling `desiredStateOfWorld.AddPlugin(...)`. What if we modified `handlePluginRegistration` in `pluginwatcher` to always call `desiredStateOfWorld.AddPlugin(...)`. And rename `desiredStateOfWorld.AddPlugin(...)` to `desiredStateOfWorld.AddOrUpdatePlugin(...)`. And then have the method update the time stamp?